src/main/java/com/artipie/npm/proxy/http/DownloadPackageSlice.java

```diff
@@ -0,0 +1,125 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.npm.proxy.http;
+
+import com.artipie.asto.Content;
+import com.artipie.http.Response;
+import com.artipie.http.Slice;
+import com.artipie.http.async.AsyncResponse;
+import com.artipie.http.rq.RequestLineFrom;
+import com.artipie.http.rs.RsStatus;
+import com.artipie.http.rs.RsWithBody;
+import com.artipie.http.rs.RsWithHeaders;
+import com.artipie.http.rs.RsWithStatus;
+import com.artipie.npm.proxy.NpmProxy;
+import com.artipie.npm.proxy.json.PackageJsonTransformer;
+import com.jcabi.log.Logger;
+import hu.akarnokd.rxjava2.interop.SingleInterop;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.StreamSupport;
+import org.cactoos.map.MapEntry;
+import org.reactivestreams.Publisher;
+
+/**
+ * HTTP slice for download package requests.
+ * @since 0.1
+ * @checkstyle ReturnCountCheck (200 lines)
+ */
+public final class DownloadPackageSlice implements Slice {
+    /**
+     * Download package request URL regexp pattern.
+     */
+    public static final Pattern PATH_PATTERN = Pattern.compile("^/(((?!/-/).)+)$");
+
+    /**
+     * NPM Proxy facade.
+     */
+    private final NpmProxy npm;
+
+    /**
+     * Ctor.
+     *
+     * @param npm NPM Proxy facade
+     */
+    public DownloadPackageSlice(final NpmProxy npm) {
+        this.npm = npm;
+    }
+
+    @Override
+    @SuppressWarnings("PMD.OnlyOneReturn")
+    public Response response(final String line,
+        final Iterable<Map.Entry<String, String>> headers,
+        final Publisher<ByteBuffer> body) {
+        Logger.debug(
+            DownloadPackageSlice.class,
+            "Processing download package request for query %s",
+            line
+        );
+        final Matcher matcher = DownloadPackageSlice.PATH_PATTERN
+            .matcher(new RequestLineFrom(line).uri().getPath());
+        if (!matcher.matches()) {
+            return RsProvider.notFound();
+        }
+        final String name = matcher.group(1);
+        Logger.debug(DownloadPackageSlice.class, "Determined package name is: %s", name);
+        return new AsyncResponse(
+            this.npm.getPackage(name)
+                .map(
+                    pkg -> (Response) new RsWithHeaders(
+                        new RsWithBody(
+                            new RsWithStatus(RsStatus.OK),
+                            new Content.From(
+                                DownloadPackageSlice.clientFormat(pkg.data(), headers).getBytes()
+                            )
+                        ),
+                        Arrays.asList(
```

Comment:
I have tried but java 8 compilation failed w/ unchecked generic array creation for varargs parameter of type java.util.Map.Entry<java.lang.String,java.lang.String>[]. It's strange because of SafeVarargs annotation is used, but it doesn't work. Reverted to Arrays.asList

Reply:
fixed w/ new ListOf<Map.Entry<String, String>>(...)